File: dependency-inversion/code.py
from abc import ABC, abstractmethod

class PaymentProcessor(ABC):
    @abstractmethod
    def pay(self, amount):
        pass

class PayPalPaymentProcessor(PaymentProcessor):
    def pay(self, amount):
        return f"Paid {amount} via PayPal"


class StripePaymentProcessor(PaymentProcessor):
    def pay(self, amount):
        return f"Paid {amount} via Stripe"


class OrderService:
    def __init__(self, payment_processor: PaymentProcessor):
        self.payment_processor = payment_processor

    def checkout(self, amount):
        return self.payment_processor.pay(amount)


paypal_processor = PayPalPaymentProcessor()
stripe_processor = StripePaymentProcessor()

order1 = OrderService(paypal_processor)
order2 = OrderService(stripe_processor)

print(order1.checkout(100)) 
print(order2.checkout(200))

File: dependency-inversion/code.js
class PaymentProcessor {
  pay(amount) {
    throw new Error("Method must be implemented");
  }
}

class PayPalPaymentProcessor extends PaymentProcessor {
  pay(amount) {
    return `Paid ${amount} via PayPal`;
  }
}

class StripePaymentProcessor extends PaymentProcessor {
  pay(amount) {
    return `Paid ${amount} via Stripe`;
  }
}

class OrderService {
  constructor(paymentProcessor) {
    this.paymentProcessor = paymentProcessor;
  }

  checkout(amount) {
    return this.paymentProcessor.pay(amount);
  }
}

const paypalProcessor = new PayPalPaymentProcessor();
const stripeProcessor = new StripePaymentProcessor();

const order1 = new OrderService(paypalProcessor);
const order2 = new OrderService(stripeProcessor);

console.log(order1.checkout(100));
console.log(order2.checkout(200));
